src/plugins/kibana/server/lib/ingest/processors/date/__tests__/converter.js

```diff
@@ -0,0 +1,119 @@
+          let expected;
+          beforeEach(function () {
+            source = {
+              processor_id: 'foo_processor_id',
+              source_field: 'foo_source_field',
+              target_field: 'foo_target_field',
+              formats: [ 'format1', 'format2' ],
+              timezone: 'foo_timezone',
+              locale: 'foo_locale'
+            };
+
+            expected = {
+              date: {
+                tag: 'foo_processor_id',
+                field: 'foo_source_field',
+                target_field: 'foo_target_field',
+                formats: [ 'format1', 'format2' ],
+                timezone: 'foo_timezone',
+                locale: 'foo_locale'
+              }
+            };
+          });
+
+          it('should convert from a kibana api object to an elasticsearch object', () => {
+            const actual = kibanaToEs(source);
+            expect(_.isEqual(actual, expected)).to.be.ok();
+          });
+
+          it('should ignore additional source fields', () => {
+            source.foo = 'bar';
+            source.bar = 'baz';
+
+            const actual = kibanaToEs(source);
+            expect(_.isEqual(actual, expected)).to.be.ok();
+          });
+
+          it('should add a custom format to the list of formats', () => {
+            source.formats.push('custom');
+            source.custom_format = 'foo_custom_format';
+            expected.date.formats.push('foo_custom_format');
+
+            const actual = kibanaToEs(source);
+            expect(_.isEqual(actual, expected)).to.be.ok();
+          });
+
+        });
+
+        describe('esToKibana', () => {
+
+          let source;
+          let expected;
+          beforeEach(function () {
+            source = {
+              tag: 'foo_tag',
+              field: 'foo_field',
+              target_field: 'foo_target_field',
+              formats: [ 'iso8601', 'unix', 'unix_ms', 'tai64n' ],
+              timezone: 'foo_timezone',
+              locale: 'foo_locale'
+            };
+
+            expected = {
+              typeId: 'date',
+              processor_id: 'foo_tag',
+              source_field: 'foo_field',
+              target_field: 'foo_target_field',
+              formats: [ 'ISO8601', 'UNIX', 'UNIX_MS', 'TAI64N' ],
+              custom_format: '',
+              timezone: 'foo_timezone',
+              locale: 'foo_locale',
+            };
+          });
+
+          it('should convert from an elasticsearch object to a kibana api object', () => {
+            const actual = esToKibana(source);
+            expect(_.isEqual(actual, expected)).to.be.ok();
+          });
+
+          it('should ignore additional source fields', () => {
+            source.foo = 'bar';
+            source.bar = 'baz';
+
+            const actual = esToKibana(source);
+            expect(_.isEqual(actual, expected)).to.be.ok();
+          });
+
+          it('should treat any unknown formats as a custom format', () => {
```

Comment:
I added logic/test to explicitly grab the last custom format specified. The UI does not support more than one custom format.